src/sql/workbench/api/node/extHostModelView.ts

```diff
@@ -34,27 +35,50 @@ class ModelBuilderImpl implements sqlops.ModelBuilder {
 
 	card(): sqlops.ComponentBuilder<sqlops.CardComponent> {
 		let id = this.getNextComponentId();
-		return new ComponentBuilderImpl(new CardWrapper(this._proxy, this._handle, id));
+		return this.withEventHandler(new CardWrapper(this._proxy, this._handle, id), id);
+	}
+
+	inputBox(): sqlops.ComponentBuilder<sqlops.InputBoxComponent> {
+		let id = this.getNextComponentId();
+		return this.withEventHandler(new InputBoxWrapper(this._proxy, this._handle, id), id);
 	}
 
 	dashboardWidget(widgetId: string): sqlops.ComponentBuilder<sqlops.WidgetComponent> {
 		let id = this.getNextComponentId();
-		return new ComponentBuilderImpl<sqlops.WidgetComponent>(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWidget, id));
+		return this.withEventHandler<sqlops.WidgetComponent>(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWidget, id), id);
 	}
 
 	dashboardWebview(webviewId: string): sqlops.ComponentBuilder<sqlops.WebviewComponent> {
 		let id = this.getNextComponentId();
-		return new ComponentBuilderImpl(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWebview, id));
+		return this.withEventHandler(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWebview, id), id);
+	}
+
+	withEventHandler<T extends sqlops.Component>(component: ComponentWrapper, id: string): sqlops.ComponentBuilder<T> {
+		let componentBuilder: ComponentBuilderImpl<T> = new ComponentBuilderImpl<T>(component);
+		this._eventHandlers.set(id, componentBuilder);
+		return componentBuilder;
+	}
+
+	handleEvent(id: string, eventArgs: IComponentEventArgs): void {
```

Comment:
We have lots of IDs. What is this one - a component ID?

Reply:
yes this is the component id

---
In reply to: [183179752](https://github.com/Microsoft/sqlopsstudio/pull/1190#discussion_r183179752) [](ancestors = 183179752)